### ISSUE TITLE
500 error on campaigns/1/opens page

### ISSUE THREAD

**Issue by fri3ndly** [spatie/mailcoach-support#101]
I am receiving a 500 error when trying to access this URL:

http://mailcoach.yourcountdown.to/campaigns/1/opens
(The clicks/unsubscribes etc work as normal)

PHP version 7.4.3
MariaDB 1:10.2.31+maria~bionic
Laravel: 6.18.1
Mailcoach : 1.8.0

Stack trace:
```
[previous exception] [object] (PDOException(code: 42000): SQLSTATE[42000]: Syntax error or access violation: 1055 'mailcoach.mailcoach_subscribers.email' isn't in GROUP BY at /home/runcloud/webapps/mailcoach/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOConnection.php:61)
[stacktrace]
#0 /home/runcloud/webapps/mailcoach/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOConnection.php(61): PDO->prepare()
#1 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Database/Connection.php(331): Doctrine\\DBAL\\Driver\\PDOConnection->prepare()
#2 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Database/Connection.php(662): Illuminate\\Database\\Connection->Illuminate\\Database\\{closure}()
#3 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Database/Connection.php(629): Illuminate\\Database\\Connection->runQueryCallback()
#4 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Database/Connection.php(338): Illuminate\\Database\\Connection->run()
#5 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2149): Illuminate\\Database\\Connection->select()
#6 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2137): Illuminate\\Database\\Query\\Builder->runSelect()
#7 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2609): Illuminate\\Database\\Query\\Builder->Illuminate\\Database\\Query\\{closure}()
#8 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2138): Illuminate\\Database\\Query\\Builder->onceWithColumns()
#9 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(537): Illuminate\\Database\\Query\\Builder->get()
#10 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(521): Illuminate\\Database\\Eloquent\\Builder->getModels()
#11 /home/runcloud/webapps/mailcoach/vendor/spatie/laravel-query-builder/src/QueryBuilder.php(64): Illuminate\\Database\\Eloquent\\Builder->get()
#12 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(721): Spatie\\QueryBuilder\\QueryBuilder->get()
#13 /home/runcloud/webapps/mailcoach/vendor/spatie/laravel-mailcoach/src/Http/App/Controllers/Campaigns/Sent/CampaignOpensController.php(16): Illuminate\\Database\\Eloquent\\Builder->paginate()
#14 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php(48): Spatie\\Mailcoach\\Http\\App\\Controllers\\Campaigns\\Sent\\CampaignOpensController->__invoke()
#15 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Routing/Route.php(219): Illuminate\\Routing\\ControllerDispatcher->dispatch()
#16 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Routing/Route.php(176): Illuminate\\Routing\\Route->runController()
#17 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Routing/Router.php(681): Illuminate\\Routing\\Route->run()
#18 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(130): Illuminate\\Routing\\Router->Illuminate\\Routing\\{closure}()
#19 /home/runcloud/webapps/mailcoach/vendor/spatie/laravel-mailcoach/src/Http/App/Middleware/SetMailcoachDefaults.php(17): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#20 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(171): Spatie\\Mailcoach\\Http\\App\\Middleware\\SetMailcoachDefaults->handle()
#21 /home/runcloud/webapps/mailcoach/vendor/spatie/laravel-mailcoach/src/Http/App/Middleware/Authorize.php(19): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#22 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(171): Spatie\\Mailcoach\\Http\\App\\Middleware\\Authorize->handle()
#23 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Auth/Middleware/Authenticate.php(43): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#24 /home/runcloud/webapps/mailcoach/vendor/spatie/laravel-mailcoach/src/Http/App/Middleware/Authenticate.php(28): Illuminate\\Auth\\Middleware\\Authenticate->handle()
#25 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(171): Spatie\\Mailcoach\\Http\\App\\Middleware\\Authenticate->handle()
#26 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php(41): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#27 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(171): Illuminate\\Routing\\Middleware\\SubstituteBindings->handle()
#28 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php(49): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#29 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(171): Illuminate\\View\\Middleware\\ShareErrorsFromSession->handle()
#30 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(56): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#31 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(171): Illuminate\\Session\\Middleware\\StartSession->handle()
#32 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php(37): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#33 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(171): Illuminate\\Cookie\\Middleware\\AddQueuedCookiesToResponse->handle()
#34 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php(66): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#35 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(171): Illuminate\\Cookie\\Middleware\\EncryptCookies->handle()
#36 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(105): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#37 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Routing/Router.php(683): Illuminate\\Pipeline\\Pipeline->then()
#38 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Routing/Router.php(658): Illuminate\\Routing\\Router->runRouteWithinStack()
#39 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Routing/Router.php(624): Illuminate\\Routing\\Router->runRoute()
#40 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Routing/Router.php(613): Illuminate\\Routing\\Router->dispatchToRoute()
#41 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(170): Illuminate\\Routing\\Router->dispatch()
#42 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(130): Illuminate\\Foundation\\Http\\Kernel->Illuminate\\Foundation\\Http\\{closure}()
#43 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#44 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(171): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle()
#45 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#46 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(171): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle()
#47 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php(27): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#48 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(171): Illuminate\\Foundation\\Http\\Middleware\\ValidatePostSize->handle()
#49 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/CheckForMaintenanceMode.php(63): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#50 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(171): Illuminate\\Foundation\\Http\\Middleware\\CheckForMaintenanceMode->handle()
#51 /home/runcloud/webapps/mailcoach/vendor/fideloper/proxy/src/TrustProxies.php(57): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#52 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(171): Fideloper\\Proxy\\TrustProxies->handle()
#53 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(105): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#54 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(145): Illuminate\\Pipeline\\Pipeline->then()
#55 /home/runcloud/webapps/mailcoach/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(110): Illuminate\\Foundation\\Http\\Kernel->sendRequestThroughRouter()
#56 /home/runcloud/webapps/mailcoach/public/index.php(55): Illuminate\\Foundation\\Http\\Kernel->handle()
#57 {main}
"} 
```

**Comment by freekmurze** [comment]
I'm thinking this is caused by a subtle difference between MySQL and Maria. If possible try to switch to MySQL. I'll try to recreate and fix this problem soon.

**Comment by raulp** [comment]
That's because of the default sql_mode for your mysql install. You can remove ONLY_FULL_GROUP_BY, but i think @freekmurze should look into how to rewrite the queries to work with ONLY_FULL_GROUP_BY. I'm going to start installing this today and maybe i can provide some help once i get to that part.

**Comment by etenzy** [comment]
The Query is:
```sql
select
    mailcoach_subscribers.id as subscriber_id,
    mailcoach_subscribers.email as subscriber_email,
    count(mailcoach_subscribers.id) as open_count,
    min(mailcoach_campaign_opens.created_at) AS first_opened_at
from `mailcoach_campaign_opens` 
inner join `mailcoach_sends` on `mailcoach_sends`.`id` = `mailcoach_campaign_opens`.`send_id`
inner join `mailcoach_campaigns` on `mailcoach_campaigns`.`id` = `mailcoach_sends`.`campaign_id`
inner join `mailcoach_subscribers` on `mailcoach_subscribers`.`id` = `mailcoach_sends`.`subscriber_id` 
where `mailcoach_campaigns`.`id` = 1 
group by `mailcoach_subscribers`.`id` 
order by `first_opened_at` desc 
limit 15 
offset 0
```

Since it’s not correct to select a column that is not part of the grouping, you could rewrite the query without those columns. But then you would need 2 more queries for selecting subscriber_email and first_opened_at.

A "better" (more compatible) solution would be to adjust the configuration of the mariadb-server:

```
#set the complete "forgiving" mode
mysql> SET GLOBAL sql_mode='';
```

OR

```
# Only drop the ONLY_FULL_GROUP_BY from the default. (The default SQL mode in MySQL 5.7)
mysql> SET GLOBAL sql_mode='STRICT_TRANS_TABLES,STRICT_ALL_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,TRADITIONAL,NO_ENGINE_SUBSTITUTION';
```

**Comment by raulp** [comment]
That is totally not "better", you can read more about ONLY_FULL_GROUP_BY and why it's the default. Without it, your queries can return inconsistent data. The solution for this is indeed to rewrite the queries, even if you have to do 1-2 extra queries. But i agree, temporary, disable ONLY_FULL_GROUP_BY until a fix is in place.

**Comment by freekmurze** [comment]
Hi @raulp,

I would appreciate some help here. If you don't find the time, we'll be sure to address this problem in the coming weeks.

**Comment by raulp** [comment]
hey @freekmurze, i just installed this on my local computer.

Let me find the code and get a hold of the code and i'll try to see if i can fix that query.
Where should i make that PR? Or how should i get back to you the changes?

Also, i wanted to ask you how to submit PR's about some changes or expanding this? Ex: i love how you have 2 different mail settings, but why not allow each campaign to send with a different SMTP provider? Let me give you an example: my company needs to send 10 campaigns with RANDOM driver. Also, what about adding more than 1 SES driver?

I have so many ideas that i would love to see, but not sure what is the best way to move forward. Would love to have a way to DM you (skype or twitter or whatever) before i start working on some of these crazy ideas, as i love your boilerplate and i always respected the way your company writes code and pacakges.

Thanks

**Comment by freekmurze** [comment]
> Let me find the code and get a hold of the code and i'll try to see if i can fix that query. Where should i make that PR? Or how should i get back to you the changes?

Currently we can't invite people to the repo without having to pay GitHub. It think it's best if you clone the repo, make the necessary changes, and create a diff file (an IDE like PhpStorm can do that), and mail me that file. 

> Also, i wanted to ask you how to submit PR's about some changes or expanding this? Ex: i love how you have 2 different mail settings, but why not allow each campaign to send with a different SMTP provider? Let me give you an example: my company needs to send 10 campaigns with RANDOM driver. Also, what about adding more than 1 SES driver?

When creating something like Mailcoach, it's a balancing act to keeping the code simple and adding features. We have to decided which features are applicable for the majority of users. Our thought process was that not many users would need to use a different mailer per campaign. We think that allowing this to be configured on the list level is enough for most. 

If you want this you could emulate this by changing the mailer of the list, right before you send a campaign. Not ideal, but it'll work.

> Also, what about adding more than 1 SES driver?

When installed as a package you can configure as much mailers as your heart desires.

> I have so many ideas that i would love to see, but not sure what is the best way to move forward. 

Great! I think creating an issue here per idea should work. This way my team and I can give you feedback on each of them.

> i always respected the way your company writes code and pacakges.

Thanks, I'm glad you like our work.

**Comment by freekmurze** [comment]
This problem is solved in v2.0.4 (thanks for the fix @raulp)
